### ISSUE TITLE
[Bug] DataBackup doesn't restore Signal (Messaging App) correctly

### ISSUE THREAD

**Issue by lxwulf** [XayahSuSuSu/Android-DataBackup#113]
**Describe the bug**
I can back up the signal app and also restoring, BUT Signal is crashing when starting it.

**To Reproduce**
1. Install Signal
2. Setup Signal to chat or restore from a signal backup
3. Start DataBackup and backup signal app with app and data
4. Delete Signal app
5. Restore Signal app with DataBackup.
6. Start Signal
7. Signal is instantly crashing.

**Expected behavior**
Signal starts without crashing and is in that state which I backed up.

**Screenshots**
I have a screenshot from logcat of the android system, I hope this helps in someway.
![image](https://user-images.githubusercontent.com/44492381/226727176-e70d5c96-e6f2-4384-9686-411429f16ca2.png)

**Smartphone (please complete the following information):**
 - Device: ASUS Zenfone 6
 - OS: Android 11
 - Version: ??? (Build Number: RKQ1.200710.002WW_Phone-18.0610.2106.156)
 - Kernel: 4.14.170-perf

**Additional context**
So far no additional context

**Comment by XayahSuSuSu** [comment]
Related: https://github.com/MuntashirAkon/AppManager/issues/82

**Comment by lxwulf** [comment]
So then I lost my data and have a useless backup, right? =/ 
Deactivating PIN/Fingerprint on the lock screen didn't help either.

**Comment by XayahSuSuSu** [comment]
> So then I lost my data and have a useless backup, right? =/ Deactivating PIN/Fingerprint on the lock screen didn't help either – which make sense because I backed up when I had it activated…
> 
> Update: I still have the keystore file from signal, could I somehow still get the backup to work? The key file is from 2022, so I believe it's the old/original one.

Nope, I need some time to have deeper research on it.
Maybe you won't lose it.

**Comment by XayahSuSuSu** [comment]
> 

Can you try [this](https://community.signalusers.org/t/howto-manual-backup-restore-if-full-backup-does-not-work/2462)?

**Comment by lxwulf** [comment]
> > So then I lost my data and have a useless backup, right? =/ Deactivating PIN/Fingerprint on the lock screen didn't help either – which make sense because I backed up when I had it activated…
> > Update: I still have the keystore file from signal, could I somehow still get the backup to work? The key file is from 2022, so I believe it's the old/original one.
> 
> Nope, I need some time to have deeper research on it. Maybe you won't lose it.

Very nice that you try to help! Really appreciate this!

**Comment by lxwulf** [comment]
> > So then I lost my data and have a useless backup, right? =/ Deactivating PIN/Fingerprint on the lock screen didn't help either – which make sense because I backed up when I had it activated…
> > Update: I still have the keystore file from signal, could I somehow still get the backup to work? The key file is from 2022, so I believe it's the old/original one.
> 
> Can you try [this](https://community.signalusers.org/t/howto-manual-backup-restore-if-full-backup-does-not-work/2462)?

Yes, somehow I tried but did not work. My problem is also that my only backup that I have is with lock screen because did not know about this bad situation that would come to me.

PS: And sorry miss clicked, because of the closing.

**Comment by XayahSuSuSu** [comment]
> > > So then I lost my data and have a useless backup, right? =/ Deactivating PIN/Fingerprint on the lock screen didn't help either – which make sense because I backed up when I had it activated…
> > > Update: I still have the keystore file from signal, could I somehow still get the backup to work? The key file is from 2022, so I believe it's the old/original one.
> > 
> > 
> > Can you try [this](https://community.signalusers.org/t/howto-manual-backup-restore-if-full-backup-does-not-work/2462)?
> 
> Yes, I already tried. Somehow it didn't work. My problem is also that my only backup that I have is with lock screen because did not know about this bad situation that would come to me.
> 
> PS: And sorry miss clicked, because of the closing.

What's ur android version

**Comment by lxwulf** [comment]
Android 11

Additional:
I also thought about to merge the backup from desktop app and the android app with the [signalbackup-tools](https://github.com/bepaald/signalbackup-tools). But the pin somehow is also not correct, but inside the app it worked as I had the app installed before my miserable situation. So my backup is my last chance to restore everything.

**Comment by XayahSuSuSu** [comment]
> So then I lost my data and have a useless backup, right? =/ Deactivating PIN/Fingerprint on the lock screen didn't help either – which make sense because I backed up when I had it activated…
> 
> Update: I still have the keystore file from signal, could I somehow still get the backup to work? The key file is from 2022, so I believe it's the old/original one.

I am on Android 12 now and the `Android Keystore` [has changed](https://stackoverflow.com/questions/71108966/andoid-12-keystore-user-0-missing-foder) since A12, so I can't test easily. 
I learned that the data was [encrypted with a secret key](https://community.signalusers.org/t/please-provide-option-to-not-store-crypto-keys-in-hardware-it-breaks-fullbackups/4962), so it may be the [only chance](https://community.signalusers.org/t/howto-manual-backup-restore-if-full-backup-does-not-work/2462) to make it if your keystore file corresponds to your backup.

**Comment by XayahSuSuSu** [comment]
>  if your keystore file corresponds to your backup.

Can you confirm that?
If you do, maybe things become easy.
https://knugi.com/202107/151300-Decrypting-Signal-Conversation-Database.html
https://rado0z.github.io/Decrypt_Android_Database

**Comment by lxwulf** [comment]
I really want to take more time for this. For me, it's not that imported anymore to decrypt my database, but still I want to help to get this working – when even possible.  When I have my own decryption key (keystore file) still on my phone, I would try to do the testing on my own database, else I would set up a new android installation with signal and your backup app.

Beside your very interesting links, are there more things I need to look at when I will do the reporting? :thinking: 

Sorry for my late answers :pray: 


I hope I have time in the next weeks, I will of course report here.

**Comment by XayahSuSuSu** [comment]
> I really want to take more time for this. For me, it's not that imported anymore to decrypt my database, but still I want to help to get this working – when even possible. When I have my own decryption key (keystore file) still on my phone, I would try to do the testing on my own database, else I would set up a new android installation with signal and your backup app.
> 
> Beside your very interesting links, are there more things I need to look at when I will do the reporting? 🤔
> 
> Sorry for my late answers 🙏
> 
> I hope I have time in the next weeks, I will of course report here.

Will temporarily add a warning for those apps having keystore while backing up.
The keystore turned into sqlite db file after Android 12, and I haven't figured out any solutions for that.
Good luck to you:), will keep this issue open for now.
